src/models.zig

```diff
@@ -149,6 +161,296 @@ pub const Model = struct {
+		defer data.deinit();
+		var dataWriter = data.writer();
+
+		try dataWriter.writeAll(vertData.items);
+		try dataWriter.writeAll(uvData.items);
+		try dataWriter.writeAll(normData.items);
+		try dataWriter.writeAll(faceData.items);
+
+		try main.files.write(path, data.items);
+	}
+
+	pub fn loadModel(data: []const u8) !u16 {
+		var vertices = std.ArrayList(Vec3f).init(main.stackAllocator.allocator);
+		defer vertices.deinit();
+
+		var normals = std.ArrayList(Vec3f).init(main.stackAllocator.allocator);
+		defer normals.deinit();
+
+		var uvs = std.ArrayList(Vec2f).init(main.stackAllocator.allocator);
+		defer uvs.deinit();
+
+		var tris = std.ArrayList(Triangle).init(main.stackAllocator.allocator);
+		defer tris.deinit();
+
+		var quadFaces = std.ArrayList(Quad).init(main.stackAllocator.allocator);
+		defer quadFaces.deinit();
+
+		var fixed_buffer = std.io.fixedBufferStream(data);
+		var buf_reader = std.io.bufferedReader(fixed_buffer.reader());
+		var in_stream = buf_reader.reader();
+		var buf: [128]u8 = undefined;
+		while (try in_stream.readUntilDelimiterOrEof(&buf, '\n')) |lineUntrimmed| {
+			if (lineUntrimmed.len < 3)
+				continue;
+			
+			var line = lineUntrimmed;
+			if (line[line.len - 1] == '\r') {
+				line = line[0..line.len - 1];
+			}
+			
+			if (std.mem.eql(u8, line[0..1], "#"))
+				continue;
+			
+			if (std.mem.eql(u8, line[0..2], "v ")) {
+				var coordsIter = std.mem.split(u8, line[2..], " ");
+				var coords: Vec3f = undefined;
+				var i: usize = 0;
+				while (coordsIter.next()) |coord| : (i += 1) {
+					coords[i] = try std.fmt.parseFloat(f32, coord);
+				}
+				const coordsCorrect: Vec3f = .{coords[0] + 0.5, -coords[2] + 0.5, coords[1]};
+				try vertices.append(coordsCorrect);
+			} else if (std.mem.eql(u8, line[0..3], "vn ")) {
+				var coordsIter = std.mem.split(u8, line[3..], " ");
+				var norm: Vec3f = undefined;
+				var i: usize = 0;
+				while (coordsIter.next()) |coord| : (i += 1) {
+					norm[i] = try std.fmt.parseFloat(f32, coord);
+				}
+				const normCorrect: Vec3f = .{norm[0], -norm[2], norm[1]};
+				try normals.append(normCorrect);
+			} else if (std.mem.eql(u8, line[0..3], "vt ")) {
+				var coordsIter = std.mem.split(u8, line[3..], " ");
+				var uv: Vec2f = undefined;
+				var i: usize = 0;
+				while (coordsIter.next()) |coord| : (i += 1) {
+					uv[i] = try std.fmt.parseFloat(f32, coord);
+				}
+				uv[0] *= 4;
+				uv[1] *= 4;
+				try uvs.append(.{uv[0], uv[1]});
+			} else if (std.mem.eql(u8, line[0..2], "f ")) {
+				if (std.mem.count(u8, line[2..], " ") + 1 == 3) {
+					var coordsIter = std.mem.split(u8, line[2..], " ");
+					var faceData: [3][3]usize = undefined;
+					var i: usize = 0;
+					while (coordsIter.next()) |vertex| : (i += 1) {
+						var d = std.mem.split(u8, vertex, "/");
+						var j: usize = 0;
+						while (d.next()) |value| : (j += 1) {
+							faceData[j][i] = try std.fmt.parseUnsigned(usize, value, 10) - 1;
```

Comment:
Here are some problems:
- `1//2` is perfectly valid, so you could fail loading these
- The vertex data may have fewer than 2 `/`, like `1/2` or just `1`. In these cases you are producing `undefined` indices, which is begging for a segmentation fault.
- What if the indices are out of bounds of the vertex/uv/normal arrays? Always sanitize user data!